Deoptionalization not detected

Under `--strict-optional`, the following yields an error:

```
f=None  # type: Optional[Callable[..., None]]

# ...

if f is None:
    f = lambda *x: None
f()  # error: None not callable
```

Ideally, `f` should be understood not to be None at that point.

I have a similar problem:

    input = "10" # this can be either "" or a string containing any long value
    value = long(input,0) if input!="" else None
    if value==None or value>=0x1000000000000000:
        pass
    if value==None or long(value)>=0x1000000000000000:
        pass
    if value==None or (value!=None and value>=0x1000000000000000):
        pass

I get the following errors:

    mypytestcase.py:3: error: Unsupported operand types for <= ("int" and "Optional[int]")
    mypytestcase.py:5: error: Argument 1 to "int" has incompatible type "Optional[int]"; expected "SupportsInt"
    mypytestcase.py:7: error: Unsupported operand types for <= ("int" and "Optional[int]")

I my opinion, all of these options should work.

Here's the root of the problem:
```py
reveal_type(lambda x: None)
```
```
test.py:1: error: Revealed type is 'Any'
```

Looking at `visit_func_expr` and `infer_lambda_type_using_context` in `checkexpr.py`, it appears that we infer `Any` for lambdas unless they have a `Callable` context.  We should be able to fix this by looking for a unique `Callable` in `Union` contexts.

But why is a `lambda` not at least inferred as `Callable`? We know as much about it as we do about an unannotated `def`. This is really strange:
```python
def f(x): return x
g = lambda x: x
reveal_type(f)
reveal_type(g)
```
Gives
```
__tmp__.py:3: error: Revealed type is 'def (x: Any) -> Any'
__tmp__.py:4: error: Revealed type is 'Any'
```
I think they should reveal the same type.

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -1772,6 +1772,12 @@ class ExpressionChecker:
         """
         # TODO also accept 'Any' context
         ctx = self.chk.type_context[-1]
+
+        if isinstance(ctx, UnionType):
+            callables = [t for t in ctx.items if isinstance(t, CallableType)]
+            if len(callables) == 1:
+                ctx = callables[0]
+
         if not ctx or not isinstance(ctx, CallableType):
             return None
 

--- a/test-data/unit/check-optional.test
+++ b/test-data/unit/check-optional.test
@@ -518,3 +518,10 @@ if x is not None:
     x.x = 1 # OK here
 
 [builtins fixtures/ops.pyi]
+
+[case testOptionalLambdaInference]
+from typing import Optional, Callable
+f = None # type: Optional[Callable[[int], None]]
+f = lambda x: None
+f(0)
+[builtins fixtures/function.pyi]

